apisix/plugins/ocsp-stapling.lua

```diff
@@ -0,0 +1,253 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one
+-- or more contributor license agreements.  See the NOTICE file
+-- distributed with this work for additional information
+-- regarding copyright ownership.  The ASF licenses this file
+-- to you under the Apache License, Version 2.0 (the
+-- "License"); you may not use this file except in compliance
+-- with the License.  You may obtain a copy of the License at
+--
+--   http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing,
+-- software distributed under the License is distributed on an
+-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+-- KIND, either express or implied.  See the License for the
+-- specific language governing permissions and limitations
+-- under the License.
+--
+
+local require = require
+local get_request = require("resty.core.base").get_request
+local http = require("resty.http")
+local ngx = ngx
+local ngx_ocsp = require("ngx.ocsp")
+local ngx_ssl = require("ngx.ssl")
+local radixtree_sni = require("apisix.ssl.router.radixtree_sni")
+local core = require("apisix.core")
+local apisix_ssl = require("apisix.ssl")
+
+local plugin_name = "ocsp-stapling"
+local ocsp_resp_cache = ngx.shared[plugin_name]
+
+local plugin_schema = {
+    type = "object",
+    properties = {},
+}
+
+local _M = {
+    name = plugin_name,
+    schema = plugin_schema,
+    version = 0.1,
+    priority = -44,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(plugin_schema, conf)
+end
+
+
+-- same as function set_pem_ssl_key() from "apisix.ssl.router.radixtree_sni"
+local function set_pem_ssl_key(sni, cert, pkey)
+    local r = get_request()
+    if r == nil then
+        return false, "no request found"
+    end
+
+    local parsed_cert, err = apisix_ssl.fetch_cert(sni, cert)
+    if not parsed_cert then
+        return false, "failed to parse PEM cert: " .. err
+    end
+
+    local ok, err = ngx_ssl.set_cert(parsed_cert)
+    if not ok then
+        return false, "failed to set PEM cert: " .. err
+    end
+
+    local parsed_pkey, err = apisix_ssl.fetch_pkey(sni, pkey)
+    if not parsed_pkey then
+        return false, "failed to parse PEM priv key: " .. err
+    end
+
+    ok, err = ngx_ssl.set_priv_key(parsed_pkey)
+    if not ok then
+        return false, "failed to set PEM priv key: " .. err
+    end
+
+    return true
+end
```

Comment:
IMHO, we should not define duplicated functions, it is better to export the function `set_pem_ssl_key` in `apisix.ssl.router.radixtree_sni` instead